### PR TITLE
MDS-2029 Feature/ Add filtering by application status code to GET /variances

### PR DESCRIPTION
**Pull request by NWCalvank**

### CODE REVIEW

File: python-backend/app/api/variances/resources/variance_resource.py [hunk 1]
```diff
@@ -18,22 +19,47 @@ class VarianceResource(Resource, UserMixin, ErrorMixin):
         description='Get a list of variances.',
         params={
             'page': f'The page number of paginated records to return. Default: {PAGE_DEFAULT}',
-            'per_page': f'The number of records to return per page. Default: {PER_PAGE_DEFAULT}'
+            'per_page': f'The number of records to return per page. Default: {PER_PAGE_DEFAULT}',
+            'variance_application_status_code':
+            'Comma-separated list of code statuses to include in results. Default: All status codes.',
```
Comment (thegentlemanphysicist) [inline python-backend/app/api/variances/resources/variance_resource.py]: is there no point to using an f-string when we don't add a {variable} ?

File: python-backend/app/api/variances/resources/variance_resource.py [hunk 2]
```diff
@@ -18,22 +19,47 @@ class VarianceResource(Resource, UserMixin, ErrorMixin):
...
+    def _apply_filters_and_pagination(self,
+                                      page_number=PAGE_DEFAULT,
+                                      page_size=PER_PAGE_DEFAULT,
+                                      application_status=None):
+        status_filter_values = list(map(
+            lambda x: x.variance_application_status_code,
+            VarianceApplicationStatusCode.active()))
+
+        if application_status is not None:
+            status_filter_values = application_status.split(',')
```
Comment (thegentlemanphysicist) [inline python-backend/app/api/variances/resources/variance_resource.py]: Is it possible to have multiple application statuses or is this used in searching for all variances in a list of application statuses?

File: python-backend/tests/status_code_gen.py [hunk 3]
```diff
@@ -95,3 +96,11 @@ def SampleDangerousOccurrenceSubparagraphs(num):
         db.session.query(ComplianceArticle).filter(
             ComplianceArticle.section == '1', ComplianceArticle.sub_section == '7',
             ComplianceArticle.paragraph == '3', ComplianceArticle.sub_paragraph != None).all(), num)
+
+def RandomVarianceApplicationStatusCode():
+    return random.choice([
+        x.variance_application_status_code
+        for x in filter(
+            lambda x: x.variance_application_status_code not in ['APP', 'DEN'],
+            VarianceApplicationStatusCode.active())
```
Comment (thegentlemanphysicist) [inline python-backend/tests/status_code_gen.py]: Is this setting the status to active when the variance is not app or denied?

File: python-backend/app/api/variances/resources/variance_resource.py (hunk 1 above)
Comment (NWCalvank) [reply python-backend/app/api/variances/resources/variance_resource.py]: That's correct 😄 the `f` allows for string interpolation, like backticks in JS

File: python-backend/app/api/variances/resources/variance_resource.py (hunk 2 above)
Comment (NWCalvank) [reply python-backend/app/api/variances/resources/variance_resource.py]: This searches across all variances and only returns the ones with the requested application status(es). If no statuses were provided, it filters with all statuses -- ie. no variances will be filtered out.

File: python-backend/tests/status_code_gen.py (hunk 3 above)
Comment (NWCalvank) [reply python-backend/tests/status_code_gen.py]: This is just using the `.active()` method, which is a custom method on the `VarianceApplicationStatusCode` model, to run a query for all active variance application status code options. Then, it filters 'APP' and 'DEC' out of the list, returning all the other codes.

The reason for filtering these out is that approved and denied variances have requirements for other column values. This is part of the tests factory code, so you can see how it's used and how APP & DEC are handled in `factories.py`

File: python-backend/app/api/variances/resources/variance_resource.py (hunk 1 above)
Comment (thegentlemanphysicist) [reply python-backend/app/api/variances/resources/variance_resource.py]: cool

**Pull request merged**